### PR TITLE
fix: harden metrics collection edge cases

### PR DESCRIPTION
**Pull request by themoah**
## Summary by Sourcery

Harden metrics collection around startup, empty cycles, and hot-partition throughput, and bump the project/chart versions.

Bug Fixes:
- Prevent metrics collector startup from failing when Kafka is unreachable by degrading the first collection cycle instead of exiting.
- Ensure an empty collection cycle still performs full cleanup and refreshes the metrics control-plane snapshot to avoid stale state.
- Fix hot-partition throughput detection so topic names containing ':' no longer break collection cycles by avoiding key-string parsing.

Enhancements:
- Expose partition throughput metrics grouped by topic from the partition throughput tracker, using the retained topic/partition metadata instead of parsing compound keys.

Build:
- Bump application version from 0.2.12 to 0.2.13 in the Gradle build.
- Bump Helm chart version from 0.3.7 to 0.3.8 and align appVersion to 0.2.13 with updated release notes.

Tests:
- Add startup and empty-cycle tests for the metrics collector to verify degraded startup and snapshot refresh behaviour when Kafka is down or has no consumer groups.
- Add hot-partition detector tests covering colon-containing topic names and cleanup behaviour to ensure throughput and lag detection remain correct.

## Summary by CodeRabbit

* **New Features**
  * Added batched Kafka offset retrieval with configurable concurrency limits.
  * Enabled virtual threads by default in the Helm chart.
  * Added support for leaderless partitions and topic names containing colons.
* **Bug Fixes**
  * Improved startup recovery when Kafka is unavailable.
  * Improved empty snapshot refreshes and consumer-group failure handling.
  * Improved partition metadata and offset fallback handling.
* **Documentation**
  * Updated configuration, troubleshooting, and release documentation.
* **Tests**
  * Added coverage for batching, startup recovery, snapshots, special topic names, and partition edge cases.

### CODE REVIEW

File: src/main/java/io/github/themoah/klag/metrics/MetricsCollector.java [hunk 1]
```diff
@@ -639,51 +563,115 @@ private void reportMetrics(
...
-        log.warn("Failed to collect lag for group {} (skipped this cycle): {}",
-          groupId, err.getMessage());
-        return Future.succeededFuture(null);
-      });
+          }
+          return null;
+        });
+      }
+    ).map(v -> byGroup);
   }
```
[the remark this answers is not in the document]
Comment (themoah) [reply src/main/java/io/github/themoah/klag/metrics/MetricsCollector.java]: Confirmed and fixed in 8c22b53. Reproduced it first: `MetricsCollectorChunkingTest.skippedGroupOffsetsDoNotDeleteThatGroupsGauges` fails on the old code (the group survives one cycle via the two-phase gauge cleanup, then gets deleted on the second consecutive failure).

`fetchGroupOffsets` now takes the `CycleState` and sets `cycle.partial = true` in the per-group recover, so `finishCycle` skips every `retainAll`. `collectGroupLags` threads it through from both call sites (`collectAllGroupsParallel` had to construct the `CycleState` before starting the lag future).

File: CLAUDE.md [hunk 2]
```diff
@@ -59,6 +59,14 @@ src/main/java/io/github/themoah/klag/
...
 ```
 
+**Collection cycle (keep this shape).** Each cycle runs in two phases: committed offsets for
+every group (in waves of `KAFKA_MAX_CONCURRENT_GROUPS`), then **one** batched
+`getLogEndOffsets(Set<String>)` for the union of their topics — one `describeTopics` plus
+three `listOffsets` for the whole set, not per topic. Lag assembly is then pure computation.
+Admin request volume must stay independent of topic count; reintroducing a per-topic fetch
+regresses it to 4 requests × every topic × every cycle.
+`MetricsCollectorBatchingTest` pins this.
+
```
[the remark this answers is not in the document]
Comment (themoah) [reply CLAUDE.md:71]: Fixed in 8c22b53. Qualified both sites: CLAUDE.md now notes that with `KAFKA_CHUNK_COUNT > 1` the topic union is split into that many batches, so it is one batched call per chunk (still not per topic — that is the invariant being pinned); the Chart.yaml release note got the same parenthetical.

File: src/main/java/io/github/themoah/klag/metrics/MetricsCollector.java [hunk 3]
```diff
@@ -203,8 +203,11 @@ public Future<Void> start() {
     log.info("Starting metrics collector with interval: {}ms, filter: {}, exclude: {}",
       intervalMs, groupFilter.includeDescription(), groupFilter.excludeDescription());
 
+    // The first cycle is recovered: a broker that is down at boot must degrade (like
+    // KafkaHealthMonitor) instead of failing verticle startup, which exits the process
+    // and crash-loops under Kubernetes. collectAndReport already logs the cause.
     return reporter.start()
-      .compose(v -> collectAndReport())
+      .compose(v -> collectAndReport().recover(err -> Future.succeededFuture()))
       .onComplete(ar -> {
```
[the remark this answers is not in the document]
Comment (themoah) [reply src/main/java/io/github/themoah/klag/metrics/MetricsCollector.java:225]: Not changing this — the unconditional recover is the fix, not an oversight.

Narrowing it to "Kafka connectivity" is not reliably expressible here: `KafkaClientServiceImpl` surfaces admin failures wrapped through Vert.x futures, so classifying transient broker unreachability apart from a config error means matching on exception text. Fail-fast on the first cycle exits the verticle and crash-loops the pod, which is exactly the boot-order dependency this PR removes (`KafkaHealthMonitor` already degrades rather than failing startup for the same reason).

A genuine startup-breaking bug is not masked: the periodic timer keeps running, every subsequent failed cycle logs `Failed to collect lag metrics` at ERROR, and `/readyz` returns 503 while Kafka is DOWN. The degraded-start signal you asked for already exists in the readiness probe.

File: .gitignore [hunk 4]
```diff
@@ -162,8 +162,9 @@ buildNumber.properties
 # Ignore Gradle GUI config
 gradle-app.setting
 
-# Avoid ignoring Gradle wrapper jar file (.jar files are usually ignored)
-!gradle-wrapper.jar
+# Gradle wrapper binaries are NOT committed (CI runs gradle directly); `./gradlew`
+# regenerates gradle/wrapper/ locally on first run.
+gradle/wrapper/
```
[the remark this answers is not in the document]
Comment (themoah) [reply .gitignore]: Obsolete — the `.gitignore` change was reverted in cfe338f, before this comment could apply to the current diff.

[the remark this answers is not in the document]
Comment (themoah) [reply .gitignore]: Obsolete for this PR — the `gradle/wrapper/` ignore rule was reverted in cfe338f, so the diff no longer touches `.gitignore`.

The underlying observation (wrapper JAR is not tracked, docs say `./gradlew`) is a pre-existing repo state, not something this PR introduces. Tracking it separately rather than expanding this diff.

[the remark this answers is not in the document]
Comment (themoah) [reply .gitignore]: The `.gitignore` change this was raised against was reverted in cfe338f, so it no longer applies to this diff. The wrapper JAR being untracked while docs say `./gradlew` predates this PR — tracking separately instead of widening scope here.
